updater/osv/osv.go

```diff
@@ -0,0 +1,801 @@
+		}
+		e.repoindex[key] = i
+	}
+	return &e.Repository[i]
+}
+
+func (e *ecs) Len() int {
+	return len(e.Vulnerability)
+}
+
+func (e *ecs) Finalize() []*claircore.Vulnerability {
+	r := make([]*claircore.Vulnerability, len(e.Vulnerability))
+	for i := range e.Vulnerability {
+		r[i] = &e.Vulnerability[i]
+	}
+	return r
+}
+
+// FromCVSS is an attempt at an implementation of the scale and formulas
+// described here: https://www.first.org/cvss/v3.1/specification-document#Qualitative-Severity-Rating-Scale
+func fromCVSS(s string) (sev claircore.Severity, err error) {
+	ms := strings.Split(strings.TrimRight(s, "/"), "/") // "m" as in "metric"
+	if !strings.HasPrefix(ms[0], "CVSS:3") {
+		return 0, fmt.Errorf("unknown label: %q", ms[0])
+	}
+	// Giant switch ahoy
+	var ns [8]float64
+	for _, m := range ms[1:] {
+		vec := strings.SplitN(m, ":", 2) // TODO(hank) go1.18: use strings.Cut
+		switch n, v := vec[0], vec[1]; n {
+		// Base metrics:
+		case `AV`:
+			const i = 0
+			switch v {
+			case `N`:
+				ns[i] = 0.85
+			case `A`:
+				ns[i] = 0.62
+			case `L`:
+				ns[i] = 0.55
+			case `P`:
+				ns[i] = 0.2
+			}
+		case `AC`:
+			const i = 1
+			switch v {
+			case `L`:
+				ns[i] = 0.77
+			case `H`:
+				ns[i] = 0.44
+			}
+		case `PR`:
+			const i = 2
+			switch v {
+			case `N`:
+				ns[i] = 0.85
+			case `L`:
+				ns[i] = 0.62 // Fixup later
+			case `H`:
+				ns[i] = 0.27 // Fixup later
+			}
+		case `UI`:
+			const i = 3
+			switch v {
+			case `N`:
+				ns[i] = 0.85
+			case `R`:
+				ns[i] = 0.62
+			}
+		case `S`:
+			const i = 4
+			// This is a cheat. Encode "changed" as 1. Not actually used in
+			// calculations, just changes the values used for other metrics.
+			switch v {
+			case `U`:
+				ns[i] = 0
+			case `C`:
+				ns[i] = 1
+			}
+		case `C`:
+			const i = 5
+			switch v {
+			case `H`:
+				ns[i] = 0.56
+			case `L`:
+				ns[i] = 0.22
+			case `N`:
+				ns[i] = 0
+			}
+		case `I`:
+			const i = 6
+			switch v {
+			case `H`:
+				ns[i] = 0.56
+			case `L`:
+				ns[i] = 0.22
+			case `N`:
+				ns[i] = 0
+			}
+		case `A`:
+			const i = 7
+			switch v {
+			case `H`:
+				ns[i] = 0.56
+			case `L`:
+				ns[i] = 0.22
+			case `N`:
+				ns[i] = 0
+			}
+			// Ignore temporal and environmental metrics.
+		}
+	}
+	changed := ns[4] != 0 // if Scope == Changed
+	if changed {
+		switch ns[ /*Privileges Required*/ 2] {
+		case 0.62:
+			ns[2] = 0.68
+		case 0.27:
+			ns[2] = 0.5
+		}
+	}
+
+	var score float64
+	iss := 1 - ((1 - ns[ /*C*/ 5]) * (1 - ns[ /*I*/ 6]) * (1 - ns[ /*A*/ 7]))
+	var imp float64
+	if changed {
+		imp = 7.52*(iss-0.029) - 3.25*math.Pow((iss-0.02), 15)
+	} else {
+		imp = iss * 6.42
+	}
+	if imp > 0 { // Score is 0 when impact is 0 or below.
+		exp := 8.22 * ns[ /*AV*/ 0] * ns[ /*AC*/ 1] * ns[ /*PR*/ 2] * ns[ /*UI*/ 3]
+		s := exp + imp
+		if changed {
+			s *= 1.08
+		}
+		s = math.Min(s, 10)
+		// Roundup function, as spec'd.
+		i := int(s * 100_000)
```

Comment:
didn't know (or had possibly forgotten) you could declare numerics like this, v.nice